pkg/controller/direct/sql/normalize.go

```diff
@@ -290,6 +303,95 @@ func normalizeAuditLogBucketRef(ctx context.Context, kube client.Reader, obj *kr
+	if obj.Spec.CloneSource == nil {
+		return "", nil
+	}
+
+	sqlInstanceRef := obj.Spec.CloneSource.SQLInstanceRef
+
+	if sqlInstanceRef.External != "" && sqlInstanceRef.Name != "" {
+		return "", fmt.Errorf("cannot specify both spec.settings.cloneSource.sqlInstanceRef.external and spec.settings.cloneSource.sqlInstanceRef.name")
+	}
+
+	if sqlInstanceRef.External != "" {
+		return sqlInstanceRef.External, nil
+	} else if sqlInstanceRef.Name != "" {
+		if sqlInstanceRef.Namespace == "" {
+			sqlInstanceRef.Namespace = obj.Namespace
+		}
+
+		key := types.NamespacedName{
+			Namespace: sqlInstanceRef.Namespace,
+			Name:      sqlInstanceRef.Name,
+		}
+
+		sqlInstance := &unstructured.Unstructured{}
+		sqlInstance.SetGroupVersionKind(krm.SQLInstanceGVK)
+		if err := kube.Get(ctx, key, sqlInstance); err != nil {
+			if apierrors.IsNotFound(err) {
+				return "", k8s.NewReferenceNotFoundError(krm.SQLInstanceGVK, key)
+			}
+			return "", fmt.Errorf("error reading referenced SQLInstance %v: %w", key, err)
+		}
+
+		sqlInstanceName, err := refs.GetResourceID(sqlInstance)
+		if err != nil {
+			return "", err
+		}
+		return sqlInstanceName, nil
+	} else {
+		return "", fmt.Errorf("must specify either spec.settings.cloneSource.sqlInstanceRef.external or spec.settings.cloneSource.sqlInstanceRef.name")
+	}
+}
+
+func normalizeSourceSQLDatabaseRefs(ctx context.Context, kube client.Reader, obj *krm.SQLInstance) ([]string, error) {
+	if obj.Spec.CloneSource == nil || obj.Spec.CloneSource.SQLDatabaseRefs == nil {
+		return nil, nil
+	}
+
+	sqlDatabaseRefs := obj.Spec.CloneSource.SQLDatabaseRefs
+	sqlDatabases := make([]string, len(sqlDatabaseRefs))
+
+	for i, sqlDatabaseRef := range sqlDatabaseRefs {
+		if sqlDatabaseRef.External != "" && sqlDatabaseRef.Name != "" {
+			return nil, fmt.Errorf("cannot specify both spec.settings.cloneSource.sqlDatabaseRefs[].external and spec.settings.cloneSource.sqlDatabaseRefs[].name")
+		}
+
+		if sqlDatabaseRef.External != "" {
+			sqlDatabases[i] = sqlDatabaseRef.External
+		} else if sqlDatabaseRef.Name != "" {
+			if sqlDatabaseRef.Namespace == "" {
+				sqlDatabaseRef.Namespace = obj.Namespace
+			}
+
+			key := types.NamespacedName{
+				Namespace: sqlDatabaseRef.Namespace,
+				Name:      sqlDatabaseRef.Name,
+			}
+
+			sqlDatabase := &unstructured.Unstructured{}
+			sqlDatabase.SetGroupVersionKind(sqldatabasev1beta1.SQLDatabaseGVK)
+			if err := kube.Get(ctx, key, sqlDatabase); err != nil {
+				if apierrors.IsNotFound(err) {
+					return nil, k8s.NewReferenceNotFoundError(sqldatabasev1beta1.SQLDatabaseGVK, key)
+				}
+				return nil, fmt.Errorf("error reading referenced SQLDatabase %v: %w", key, err)
+			}
+
+			sqlDatabaseName, err := refs.GetResourceID(sqlDatabase)
+			if err != nil {
+				return nil, err
+			}
+			sqlDatabases[i] = sqlDatabaseName
+		} else {
+			return nil, fmt.Errorf("must specify either spec.settings.cloneSource.sqlDatabaseRefs[].external or spec.settings.cloneSource.sqlDatabaseRefs[].name")
```

Comment:
FYI, k8s has fieldpath to help with this problem (writing annoyingly long field paths!)

Reply:
Can you share a link to an example of this?